KKUMI-63 authorization

eekrwl opened the pull request:
## 구현내용
### 1. 로그인이 필요한 api 구분
아래와 같이 로그인이 필요한 api는 @Login 어노테이션으로 구분해줬습니다. User에는 토큰에서 가져온 유저 객체를 넣어주도록 했습니다.
```java
@PostMapping("/posts")
    public ResponseEntity<RegisterPostResponse> registerPost(@Login User user, @RequestBody RegisterPostRequest request) {
        Long savedPostId = postService.registerPost(user, request.getSubCategoryId(), request.getContent(), request.getImages());
        return ResponseEntity.ok().body(RegisterPostResponse.of(savedPostId));
    }
```

### 1-1. 위의 방식에서 변경
interceptor를 만들어 로그인이 필요한 api를 구분하는 역할을 분리해줬습니다.
@RequiresLogin 어노테이션을 달면 로그인된 사용자만 접근할 수 있습니다.
```java
@RequiresLogin
    @GetMapping("/users/me")
    public ResponseEntity<MeResponse> getMe(@Login User user) {
        MeResponse meResponse = userService.getMe(user);
        return ResponseEntity.ok(meResponse);
    }
```

## 고민사항
### 1. 로그인 후 이용가능한 api 구분을 ArgumentResolver에서 해도 되는가?
새로 @RequiresLogin과 같은 어노테이션을 만들어서 인터셉터 같은 곳에서 걸러주려고 하다가, 기존의 @Login 도 비슷한 역할이 아닌가 해서 아래와 같이 argument resolver의 끝에서 처리를 해줬습니다. 분리해주는 것이 좋을지 궁금합니다.
만약 새로 어노테이션과 인터셉터를 만들어서 구분하게 된다면 조금 중복되는 파일을 더 만들어야 될 것 같지만 각자의 역할이 명확해질 것 같긴합니다
```java
@Override
    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, WebDataBinderFactory binderFactory) throws Exception {

        HttpServletRequest request = webRequest.getNativeRequest(HttpServletRequest.class);
        String authorization = request.getHeader("Authorization");
        if (authorization != null && authorization.startsWith("Bearer ")) {
            String token = authorization.substring(7);
            if(tokenService.isValidToken(token)) { //유효한 토큰인지 검증
                UUID userUuid = tokenService.getUserUuidFromToken(token);
                User user = userService.getUserByUuid(userUuid);
                return user;
            }
        }
        throw new CommonException(ErrorCode.AUTHENTICATION_REQUIRED, "로그인 후 이용 가능합니다.", "로그인 후 이용 가능합니다.");
    }
```

### 1-1. Interceptor로 분리
로그인된 유저 정보를 주입해주는 부분과, 로그인된 유저만 사용할 수 있는 api를 구분하는 부분을 분리해줬습니다.
기존에는 argument resolver에서 모두 처리하였으나, 로그인된 유저만 사용할 수 있다는 표시를 @RequiresLogin으로 해줬습니다.

eekrwl committed:
KKUMI-63 [FEAT] #43 로그인 필요한 api 구분하는 기능 추가

새로운 어노테이션을 만들어볼까 하다가 기존의 argument resolver에서 일단 처리

eekrwl committed:
KKUMI-63 [REFACTOR] #43 기존 컨트롤러에서 로그인한 유저인지 확인하는 코드 삭제

eekrwl committed:
KKUMI-63 [SECURITY] presigned url 로그인한 사용자만 접근 가능

babjo commented:
> 새로 @RequiresLogin과 같은 어노테이션을 만들어서 인터셉터 같은 곳에서 걸러주려고 하다가, 기존의 @Login 도 비슷한 역할이 아닌가 해서 아래와 같이 argument resolver의 끝에서 처리를 해줬습니다. 분리해주는 것이 좋을지 궁금합니다.
만약 새로 어노테이션과 인터셉터를 만들어서 구분하게 된다면 조금 중복되는 파일을 더 만들어야 될 것 같지만 각자의 역할이 명확해질 것 같긴합니다

넵 resovleArgument 역할 안에서 인증까지하는건 더 많은 책임을 가지는듯합니다. 별도 필터로 분리하는게 좋을거같습니다.

eekrwl commented:
> > 새로 @RequiresLogin과 같은 어노테이션을 만들어서 인터셉터 같은 곳에서 걸러주려고 하다가, 기존의 @Login 도 비슷한 역할이 아닌가 해서 아래와 같이 argument resolver의 끝에서 처리를 해줬습니다. 분리해주는 것이 좋을지 궁금합니다.
> > 만약 새로 어노테이션과 인터셉터를 만들어서 구분하게 된다면 조금 중복되는 파일을 더 만들어야 될 것 같지만 각자의 역할이 명확해질 것 같긴합니다
> 
> 넵 resovleArgument 역할 안에서 인증까지하는건 더 많은 책임을 가지는듯합니다. 별도 필터로 분리하는게 좋을거같습니다.

네 분리하도록 하겠습니다!

eekrwl committed:
KKUMI-63 [REFACTOR] #43 로그인 필요여부 인터셉터에서 처리하도록 수정

기존: argument resolver에서 처리
변경후: 인터셉터로 따로 분리
argument resolver가 너무 많은 역할을 하는 것 같아 분리했습니다.

eekrwl committed:
KKUMI-63 [CHORE] #43 annotation 패키지 만들어서 어노테이션 이동

eekrwl committed:
KKUMI-63 [REFACTOR] #43 로그인 필수인 api에 @RequiresLogin  추가
